servers/rendering/renderer_rd/shaders/scene_forward_lights_inc.glsl

```diff
@@ -1,14 +1,34 @@
 // Functions related to lighting
 
+/* [Drobot2014a] Low Level Optimizations for GCN. */
+float sqrt_IEEE_int_approximation(float v) {
+	return intBitsToFloat(0x1fbd1df5 + (floatBitsToInt(v) >> 1));
+}
+
+vec2 sqrt_IEEE_int_approximation(vec2 v) {
+	return intBitsToFloat(0x1fbd1df5 + (floatBitsToInt(v) >> 1));
+}
+
 float D_GGX(float cos_theta_m, float alpha) {
 	float a = cos_theta_m * alpha;
 	float k = alpha / (1.0 - cos_theta_m * cos_theta_m + a * a);
 	return k * k * (1.0 / M_PI);
 }
 
-// From Earl Hammon, Jr. "PBR Diffuse Lighting for GGX+Smith Microsurfaces" https://www.gdcvault.com/play/1024478/PBR-Diffuse-Lighting-for-GGX
+// From "Course Notes: Moving Frostbite to PBR", page 12: https://media.contentapi.ea.com/content/dam/eacom/frostbite/files/course-notes-moving-frostbite-to-pbr-v32.pdf
 float V_GGX(float NdotL, float NdotV, float alpha) {
-	return 0.5 / mix(2.0 * NdotL * NdotV, NdotL + NdotV, alpha);
+	// Original formulation of G_SmithGGX Correlated
+	// lambda_v = ( -1 + sqrt ( alphaG2 * (1 - NdotL2 ) / NdotL2 + 1) ) * 0.5 f ;
+	// lambda_l = ( -1 + sqrt ( alphaG2 * (1 - NdotV2 ) / NdotV2 + 1) ) * 0.5 f ;
+	// G_ SmithGGXCorrelated = 1 / (1 + lambda_v + lambda_l ) ;
+	// V_ SmithGGXCorrelated = G_SmithGGXCorrelated / (4.0 f * NdotL * NdotV ) ;
+
+	// This is the optimized version
+	// Caution: the " NdotL *" and " NdotV *" are explicitly inversed, this is not a mistake.
+	float Lambda_GGXV = NdotL * sqrt_IEEE_int_approximation((-NdotV * alpha + NdotV) * NdotV + alpha);
+	float Lambda_GGXL = NdotV * sqrt_IEEE_int_approximation((-NdotL * alpha + NdotL) * NdotL + alpha);
+
+	return 0.5 / (Lambda_GGXV + Lambda_GGXL);
```

Comment:
This looks like it uses way more instructions than the old version, are you sure that this is faster?

Reply:
Hey, sorry to not have shared benchmarks in the meantime, hope I can today.

As for the V_GGX implementation, you're correct, it wouldn't make much sense if it were. To be honest, I started out by trying to use the version you see and only then I tried to approximate the sqrt. Even though this is backwards in their commit order.

This V_GGX implementation might be hurting the sqrt approximation gains, if anything. The original V_GGX implementation seems to have been pretty well thought out, and more performant.

As promised I'll provide further testing today.

I've disassembled the shader on my end to check on this, resulting in the following:
[V_SmithGGXCorrelated_spirv_comp.txt](https://github.com/godotengine/godot/files/15386860/V_SmithGGXCorrelated_spirv_comp.txt)

Reply:
I think the focus of the PR should be on the main change, the rewrite of the algorithm, or split it into two separate ones if the sqrt changes can be independent

Reply:
Thanks for taking a look, indeed we end up with a lot more instructions. 

I suspect that in the short term we will split this file in two and have a separate implementation for mobile and desktop. These variants that you have added are probably more appropriate for desktop as we want maximum quality on desktop (a couple instructions won't cost much). But on mobile we absolutely need to reduce the instruction count as much as possible